libpermian/plugins/anaconda_webui/__init__.py

```diff
@@ -287,17 +263,37 @@ def execute(self):
         self.log('Running: ' + ' '.join(cmd))
 
         test_env = {'WEBUI_TEST_DIR': os.path.abspath(os.path.join(self.webui_dir, 'test'))}
+        test_output = self.crc.openLogfile('output.txt', 'w', True)
 
         time.sleep(10) # Workaround, there is a race-condition, where WebUI is accessible but /run/anaconda/bus.address doesn't exist yet
         if self.use_container:
             cont_cwd = self.test_workdir.replace(self.temp_dir, '/root/workdir')
             test_env['WEBUI_TEST_DIR'] = test_env['WEBUI_TEST_DIR'].replace(self.temp_dir, '/root/workdir')
-            process = self.container.exec(cmd, self.temp_dir, env=test_env, cwd=cont_cwd, log_error=False)
+            process = self.container.popen(cmd, self.temp_dir, env=test_env, cwd=cont_cwd, stdout=test_output)
         else:
-            process = subprocess.run(cmd, cwd=self.test_workdir, env=test_env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+            process = subprocess.Popen(cmd, cwd=self.test_workdir, env=test_env, stdout=test_output, stderr=subprocess.STDOUT)
 
-        self.log(process.stdout.decode(), 'output')
-        self.log('Test finished', show=True)
+        # Wait for test to finish
+        test_timeout = time.time() + self.test_timeout
+
+        while time.time() < test_timeout:
+            try:
+                process.wait(1)
+                self.log('Test finished', show=True)
+                break
+            except subprocess.TimeoutExpired:
+                self._collect_test_logs()
+        else:
+            # Test didn't end before timeout
+            process.terminate()
+            try:
+                process.wait(10)
+            except subprocess.TimeoutExpired:
+                # Kill if it is still running
+                process.kill()
+                if process.poll() is None:
+                    LOGGER.error('Test didn\'t end before timeout, and couldn\'t be killed')
```

Comment:
This may not be necessarily true. Even in case of SIGKILL, it can take some time for the process to end. If the process is performing uninterruptible syscall, it is not terminated until such syscall ends - which may take some CPU cycles. If the process was immediately checked in such case, it would be still running but could be terminated just few miliseconds (or even seconds) later.